### ISSUE TITLE
General::checkFile() produces "false negatives"

### ISSUE THREAD

**Issue by michael-e** [symphonycms/symphony-2#2618]
When attemtping to update the JIT Image Manipulation extension to a recent version, I received the error: `.htaccess exists but is not writable.`. Looking into it, I noticed that the extension uses `General::checkFile()` to see if the `.htaccess` file is writable.

There we find:

``` php
if (
    (!is_writable($dir) || !is_readable($dir)) // Folder
    || (file_exists($file) && (!is_readable($file) || !is_writable($file))) // File
) {
    return false;
}
```

Now if a file already exists, it **will** be writable even if the directory is readable but not writable. The above, however, will return false because `(!is_writable($dir) || !is_readable($dir))` will be false.

In short: The logic _seems_ broken.

We might change this to only look into the directory if the file does not exist. Something like:

``` php
if (
    (!file_exists($file) && (!is_writable($dir) || !is_readable($dir))) // Folder
    || (file_exists($file) && (!is_readable($file) || !is_writable($file))) // File
) {
    return false;
}
```

But on the other hand the result of this might be surprising if you attempt to delete a file without write permissions on the directory.

So the main issue I see here is that the function's name is so weak and misleading. `checkFile` does not say what it does.

The current version checks if a file and its directory are readable and writable. The latter will not be needed in order to write to existing files. So you will have **false negatives**.

My proposal checks if the directory is readable and writable if the file does not exist (so you will be able to create the file if the function returns `true`). If the file exists, it will only check the file, not the directory, which is enough to know that you can write to the file — but you may not be able to delete it. So you will have **false positives**.

I don't say which one is better (see above) – I just say it is hard to decide and/or understand it because of the function's name.

Just thinking loud… On the long run, shouldn't we use more than one function checking a file? Maybe:
- General::checkFileRead()
- General::checkFileWrite()
- General::checkFileDelete()

**Comment by michael-e** [comment]
To make this more clear (from the top of my head, I hope that I will not be proven wrong):

* Creating a file **requires** write permissions on the directory.
* Writing to an existing file **does not require** write permissions on the directory.
* Deleting a file **requires** write permissions on the directory.

**Comment by designermonkey** [comment]
Those assertions are correct.

What is needed to counter the file permissions errors, read and write errors and other issues relating to the filesystem that have dogged Symphony over the years is proper abstraction of the filesystem: Proper functions that do what they say.

I know there is a broad range of functions in PHP to access and manipulate the filesystem, but hiding these behind a proper interface is the best plan for the future.

**Comment by brendo** [comment]
> but hiding these behind a proper interface is the best plan for the future.

Yeah, I'd say [Flysystem](http://flysystem.thephpleague.com/) seems to be the defacto these days.

**Comment by designermonkey** [comment]
There isn't that many places to update that I can see (compared to changing something else more substantial). The issue I see is backwards compatibility and the fact everything is static still :(

**Comment by nitriques** [comment]
@designermonkey New PHP modules still to continue to go in the static route. I still like it a lot, the api is simple to use, no need to worry about creating the object. The bad thing is static state.

@michael-e I totally agree with your proposal. Want to send a PR ?

@brendo Flysystem is really cool. I'll try to poke at it.

Thanks!

**Comment by michael-e** [comment]
Do you mean my [code proposal](#issue-171009374) above? As I said:

> I don't say which one is better (see above) – I just say it is hard to decide and/or understand it

But if you like this proposal, I can send a PR for this small change.

**Comment by nitriques** [comment]
@michael-e  I meant creating those 3 methods.

- General::checkFileRead()
- General::checkFileWrite()
- General::checkFileDelete()

But I think that

> Deleting a file requires write permissions on the directory.

You need to be able to also write the file to delete it ?

**Comment by michael-e** [comment]
AFAIK, you don't need write permissions on a file in order to delete it. You only need execute+write permissions on the parent directory.

Sorry, but if you need a PR within a few days, I won't be able to make it. Later? Hopefully.

**Comment by nitriques** [comment]
@michael-e days does not mather, I have plenty of other things to do haha ;) still at 36 unread notifications on github. But most of next week should be dedicated to Symphony 2.7. This week was mostly extensions update.

**Comment by nitriques** [comment]
@michael-e I have a POC commit in my branch ^^^^^ @designermonkey @brendo want to review this starting point ?

**Comment by nitriques** [comment]
Note to self: add a note about the deprecation and fix Michael's name.

**Comment by michael-e** [comment]
Apart from the typos (see inline comments) this looks good to me.

**Comment by nitriques** [comment]
Great, will send a PR soon.

**Comment by nitriques** [comment]
Fixed via #2684 and #2683
